In [3]:
import os
import json

# 设置数据根目录路径
# 假设脚本在仓库根目录下运行，数据在 assets/data
ROOT_DIR = os.path.join("../assets", "data")

def get_max_size(puzzles_dict):
    """
    遍历 puzzles 字典，解析第一行尺寸，返回最大尺寸字符串(按面积比较)。
    如果无法获取，返回 "-"
    """
    max_area = -1
    max_size_str = "-"

    for key, val in puzzles_dict.items():
        if not isinstance(val, dict):
            continue
            
        problem_str = val.get("problem", "")
        if not problem_str:
            continue

        # 获取第一行
        first_line = problem_str.strip().split('\n')[0].strip()
        tokens = first_line.split()

        # 尝试读取前两个数字
        if len(tokens) >= 2:
            try:
                # 假设格式为 "Width Height" 或 "Rows Cols"
                # 这里只作为数字解析计算面积，不做具体宽高定义的区分
                dim1 = int(tokens[0])
                dim2 = int(tokens[1])
                area = dim1 * dim2
                
                if area > max_area:
                    max_area = area
                    max_size_str = f"{dim1}x{dim2}"
            except ValueError:
                continue
    
    return max_size_str

def generate_markdown_table():
    if not os.path.exists(ROOT_DIR):
        print(f"Error: Directory '{ROOT_DIR}' not found.")
        return

    # 获取所有子文件夹并排序
    subdirs = [d for d in os.listdir(ROOT_DIR) if os.path.isdir(os.path.join(ROOT_DIR, d))]
    subdirs.sort() # 按字母顺序排序

    table_data = [] # 存储每一行的数据
    total_problems = 0
    total_solutions = 0

    # 表头
    headers = ["No.", "Puzzle Name", "Problems", "Solutions", "Max Size"]
    
    # 遍历每个谜题文件夹
    for idx, puzzle_name in enumerate(subdirs, 1):
        puzzle_dir = os.path.join(ROOT_DIR, puzzle_name)
        
        # 构建文件路径
        prob_json_path = os.path.join(puzzle_dir, "problems", f"{puzzle_name}_puzzles.json")
        sol_json_path = os.path.join(puzzle_dir, "solutions", f"{puzzle_name}_solutions.json")

        # 初始化当前行的变量
        p_count = "-"
        s_count = "-"
        max_size = "-"

        # 1. 处理 Problems JSON
        if os.path.exists(prob_json_path):
            try:
                with open(prob_json_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    # 统计 puzzle 数量 (基于字典 keys 数量，比 info count 更准确)
                    puzzles = data.get("puzzles", {})
                    count = len(puzzles)
                    p_count = count
                    total_problems += count
                    try:
                    # 计算最大尺寸
                        max_size = get_max_size(puzzles)
                    except Exception as e:
                        # 解析出错保持 "-"
                        pass
            except Exception as e:
                # 解析出错保持 "-"
                
                pass

        # 2. 处理 Solutions JSON
        if os.path.exists(sol_json_path):
            try:
                with open(sol_json_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    # 统计 solution 数量
                    solutions = data.get("solutions", {})
                    count = len(solutions)
                    s_count = count
                    total_solutions += count
            except Exception as e:
                pass

        table_data.append([str(idx), puzzle_name, str(p_count), str(s_count), max_size])

    # --- 生成 Markdown 输出 ---

    # 打印表头
    print(f"| {' | '.join(headers)} |")
    print(f"| {' | '.join(['---'] * len(headers))} |")

    # 打印数据行
    for row in table_data:
        print(f"| {' | '.join(row)} |")

    # 打印统计行
    # No. 为空, Name 为 Total, Counts 加粗, Size 为空
    print(f"| | **Total** | **{total_problems}** | **{total_solutions}** | - |")

if __name__ == "__main__":
    generate_markdown_table()

| No. | Puzzle Name | Problems | Solutions | Max Size |
| --- | --- | --- | --- | --- |
| 1 | ABCEndView | 607 | 607 | 8x8 |
| 2 | Akari | 970 | 970 | 100x100 |
| 3 | Battleship | - | - | - |
| 4 | Binairo | 380 | 380 | 14x14 |
| 5 | Bosanowa | 38 | 38 | 11x16 |
| 6 | Buraitoraito | 101 | 100 | 15x15 |
| 7 | Burokku | 270 | 270 | 10x10 |
| 8 | ButterflySudoku | 77 | 77 | 12x12 |
| 9 | Clueless1Sudoku | 29 | 29 | 27x27 |
| 10 | Clueless2Sudoku | 40 | 40 | 27x27 |
| 11 | Creek | 440 | 440 | 40x50 |
| 12 | Dominos | 580 | 579 | 10x11 |
| 13 | DoubleMinesweeper | - | - | - |
| 14 | EntryExit | 170 | 170 | 16x16 |
| 15 | Eulero | 290 | 290 | 5x5 |
| 16 | EvenOddSudoku | 129 | 129 | 9x9 |
| 17 | Fillomino | 840 | 840 | 50x64 |
| 18 | Fivecells | - | - | - |
| 19 | Fobidoshi | 250 | 250 | 12x12 |
| 20 | Foseruzu | - | - | - |
| 21 | Fuzuli | 160 | 160 | 8x8 |
| 22 | Gappy | 429 | 427 | 18x18 |
| 23 | Gattai8Sudoku | 120 | 120 | 21x33 |
| 24 | GokigenNaname | - | - | - |
| 25 | GrandTour | 350

| No. | Puzzle Name | Problems | Solutions | Max Size |
| --- | --- | --- | --- | --- |
| 1 | ABCEndView | 607 | 607 | 8x8 |
| 2 | Akari | 970 | 970 | 100x100 |
| 3 | Battleship | - | - | - |
| 4 | Binairo | 380 | 380 | 14x14 |
| 5 | Bosanowa | 38 | 38 | 11x16 |
| 6 | Buraitoraito | 101 | 100 | 15x15 |
| 7 | Burokku | 270 | 270 | 10x10 |
| 8 | ButterflySudoku | 77 | 77 | 12x12 |
| 9 | Clueless1Sudoku | 29 | 29 | 27x27 |
| 10 | Clueless2Sudoku | 40 | 40 | 27x27 |
| 11 | Creek | 440 | 440 | 40x50 |
| 12 | Dominos | 580 | 579 | 10x11 |
| 13 | DoubleMinesweeper | - | - | - |
| 14 | EntryExit | 170 | 170 | 16x16 |
| 15 | Eulero | 290 | 290 | 5x5 |
| 16 | EvenOddSudoku | 129 | 129 | 9x9 |
| 17 | Fillomino | 840 | 840 | 50x64 |
| 18 | Fivecells | - | - | - |
| 19 | Fobidoshi | 250 | 250 | 12x12 |
| 20 | Foseruzu | - | - | - |
| 21 | Fuzuli | 160 | 160 | 8x8 |
| 22 | Gappy | 429 | 427 | 18x18 |
| 23 | Gattai8Sudoku | 120 | 120 | 21x33 |
| 24 | GokigenNaname | - | - | - |
| 25 | GrandTour | 350 | 350 | 15x15 |
| 26 | Hakyuu | 480 | 480 | 30x45 |
| 27 | Heyawake | - | - | - |
| 28 | Hitori | 421 | 421 | 17x17 |
| 29 | JigsawSudoku | 680 | 680 | 9x9 |
| 30 | Kakkuru | 400 | 400 | 9x9 |
| 31 | Kakurasu | 280 | 280 | 11x11 |
| 32 | Kakuro | 999 | 999 | 31x46 |
| 33 | KillerSudoku | 810 | 810 | 9x9 |
| 34 | Kuromasu | 560 | 560 | 31x45 |
| 35 | Kuroshuto | - | - | - |
| 36 | LITS | 419 | 419 | 40x57 |
| 37 | Linesweeper | 310 | 310 | 16x16 |
| 38 | Magnetic | 439 | 439 | 12x12 |
| 39 | Makaro | 190 | 190 | 15x15 |
| 40 | Maze-a-pix | - | - | - |
| 41 | Minesweeper | 360 | 360 | 14x24 |
| 42 | Mosaic | 165 | 104 | 118x100 |
| 43 | Munraito | 360 | 360 | 12x12 |
| 44 | Nanbaboru | 270 | 270 | 9x9 |
| 45 | Nondango | 110 | 110 | 14x14 |
| 46 | Nonogram | 2340 | 2339 | 30x40 |
| 47 | Norinori | 289 | 289 | 36x54 |
| 48 | NumberCross | 170 | 170 | 8x8 |
| 49 | NumberSnake | 70 | 70 | 10x10 |
| 50 | OneToX | 58 | 58 | 10x10 |
| 51 | Patchwork | 211 | 211 | 12x12 |
| 52 | Pfeilzahlen | 360 | 360 | 8x8 |
| 53 | Pills | 164 | 163 | 10x10 |
| 54 | Polyiamond | - | - | - |
| 55 | Polyminoes | - | - | - |
| 56 | Renban | 150 | 150 | 9x9 |
| 57 | SamuraiSudoku | 272 | 272 | 21x21 |
| 58 | Shikaku | 500 | 500 | 31x45 |
| 59 | ShogunSudoku | 90 | 90 | 21x45 |
| 60 | Simpleloop | 70 | 70 | 17x18 |
| 61 | Skyscraper | 470 | 470 | 8x8 |
| 62 | Slitherlink | 1176 | 1153 | 60x60 |
| 63 | Snake | 230 | 230 | 12x12 |
| 64 | SoheiSudoku | 120 | 120 | 21x21 |
| 65 | SquareO | 120 | 80 | 15x15 |
| 66 | Starbattle | 307 | 307 | 15x15 |
| 67 | Sternenhimmel | 29 | 29 | 17x17 |
| 68 | Str8t | 560 | 560 | 9x9 |
| 69 | Sudoku | 125 | 125 | 16x16 |
| 70 | Suguru | 200 | 200 | 10x10 |
| 71 | SumoSudoku | 110 | 110 | 33x33 |
| 72 | Tairupeinto | 268 | 268 | 15x15 |
| 73 | Tatamibari | 150 | 150 | 14x14 |
| 74 | TennerGrid | 375 | 374 | 6x10 |
| 75 | Tent | 706 | 706 | 30x30 |
| 76 | TerraX | 80 | 80 | 17x17 |
| 77 | Thermometer | 250 | 250 | 10x10 |
| 78 | TilePaint | 109 | 109 | 16x16 |
| 79 | Trinairo | 60 | 60 | 12x12 |
| 80 | WindmillSudoku | 150 | 150 | 21x21 |
| 81 | Yajilin | 610 | 610 | 39x57 |
| 82 | YinYang | 170 | 170 | 14x14 |
| 83 | Yonmasu | 120 | 120 | 10x10 |
| 84 | monotone | - | - | - |
| | **Total** | **24762** | **24631** | - |